File: client/internal/core/jsonroundtrip_test.go
package core

import (
	"encoding/json"
	"os"
	"path/filepath"
	"reflect"
	"strings"
	"testing"
)

// Cross-language golden-fixture round-trip on the Go side: each fixture under
// docs/contracts/fixtures/ is unmarshaled into its typed core struct, re-marshaled, and the result
// is asserted semantically equal to the original fixture. The identical files are round-tripped by
// the Java JsonRoundTripTest; both suites green ⇒ Go and Java agree on the wire contract (issue #3).
//
// Comparison is semantic: both JSON blobs are decoded into generic any trees (maps are
// order-independent) and compared with reflect.DeepEqual, the Go analogue of the Java side's
// JsonNode equality. Field names, casing, types and null-omission must match — that is the point.

// fixturesDir locates docs/contracts/fixtures/ relative to this package (client/internal/core).
func fixturesDir(t *testing.T) string {
	t.Helper()
	// up from internal/core -> internal -> client -> repo root
	dir, err := filepath.Abs(filepath.Join("..", "..", "..", "docs", "contracts", "fixtures"))
	if err != nil {
		t.Fatalf("resolve fixtures dir: %v", err)
	}
	if _, err := os.Stat(dir); err != nil {
		t.Fatalf("fixtures dir %s not found: %v", dir, err)
	}
	return dir
}

func readFixture(t *testing.T, name string) []byte {
	t.Helper()
	b, err := os.ReadFile(filepath.Join(fixturesDir(t), name))
	if err != nil {
		t.Fatalf("read fixture %s: %v", name, err)
	}
	return b
}

// decodeTree decodes raw JSON into a generic, order-independent tree for semantic comparison.
func decodeTree(t *testing.T, raw []byte) any {
	t.Helper()
	var v any
	if err := json.Unmarshal(raw, &v); err != nil {
		t.Fatalf("decode tree: %v\nraw: %s", err, raw)
	}
	return v
}

// assertRoundTrips unmarshals the fixture into *dst, re-marshals it, and asserts the re-marshaled
// JSON is semantically equal to the original fixture.
func assertRoundTrips(t *testing.T, fixtureName string, dst any) {
	t.Helper()
	original := readFixture(t, fixtureName)
	if err := json.Unmarshal(original, dst); err != nil {
		t.Fatalf("unmarshal %s into %T: %v", fixtureName, dst, err)
	}
	reserialized, err := json.Marshal(dst)
	if err != nil {
		t.Fatalf("marshal %T: %v", dst, err)
	}
	want := decodeTree(t, original)
	got := decodeTree(t, reserialized)
	if !reflect.DeepEqual(want, got) {
		t.Fatalf("round-trip of %s differs:\n want: %s\n  got: %s", fixtureName, original, reserialized)
	}
}

func TestIdentityPageRoundTrips(t *testing.T) { assertRoundTrips(t, "identity_page.json", &Identity{}) }
func TestIdentityProjectRoundTrips(t *testing.T) {
	assertRoundTrips(t, "identity_project.json", &Identity{})
}
func TestPageRoundTrips(t *testing.T)        { assertRoundTrips(t, "page.json", &Page{}) }
func TestObservationRoundTrips(t *testing.T) { assertRoundTrips(t, "observation.json", &Observation{}) }
func TestObservationMinimalRoundTrips(t *testing.T) {
	assertRoundTrips(t, "observation_minimal.json", &Observation{})
}
func TestSessionRoundTrips(t *testing.T)      { assertRoundTrips(t, "session.json", &Session{}) }
func TestSessionOpenRoundTrips(t *testing.T)  { assertRoundTrips(t, "session_open.json", &Session{}) }
func TestLinkRoundTrips(t *testing.T)         { assertRoundTrips(t, "link.json", &Link{}) }
func TestLinkDeferredRoundTrips(t *testing.T) { assertRoundTrips(t, "link_deferred.json", &Link{}) }
func TestHandoffRoundTrips(t *testing.T)      { assertRoundTrips(t, "handoff.json", &Handoff{}) }
func TestHandoffOpenRoundTrips(t *testing.T)  { assertRoundTrips(t, "handoff_open.json", &Handoff{}) }

func TestObservationKindsFixtureMatchesEnum(t *testing.T) {
	var tokens []string
	if err := json.Unmarshal(readFixture(t, "observation_kinds.json"), &tokens); err != nil {
		t.Fatalf("unmarshal kinds: %v", err)
	}
	if len(tokens) != len(ObservationKinds) {
		t.Fatalf("kind count mismatch: fixture %d vs enum %d", len(tokens), len(ObservationKinds))
	}
	for i, tok := range tokens {
		if ObservationKind(tok) != ObservationKinds[i] {
			t.Errorf("kind[%d]: fixture %q != enum %q", i, tok, ObservationKinds[i])
		}
		if ParseObservationKind(tok) != ObservationKinds[i] {
			t.Errorf("ParseObservationKind(%q) != %q", tok, ObservationKinds[i])
		}
	}
}

// TestHandoffEmptyListsSerializeAsArray guards the no-`omitempty` decision: an open handoff with
// empty lists must emit "openQuestions":[] / "nextSteps":[], never null or omitted, to match the
// server. (The fixture round-trip already covers this, but this asserts the marshaled bytes.)
func TestHandoffEmptyListsSerializeAsArray(t *testing.T) {
	var h Handoff
	if err := json.Unmarshal(readFixture(t, "handoff_open.json"), &h); err != nil {
		t.Fatalf("unmarshal: %v", err)
	}
	b, err := json.Marshal(h)
	if err != nil {
		t.Fatalf("marshal: %v", err)
	}
	s := string(b)
	for _, want := range []string{`"openQuestions":[]`, `"nextSteps":[]`} {
		if !strings.Contains(s, want) {
			t.Errorf("expected %s in %s", want, s)
		}
	}
}

File: server/src/test/java/com/agentmemory/core/JsonRoundTripTest.java
package com.agentmemory.core;

import static org.assertj.core.api.Assertions.assertThat;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;
import org.junit.jupiter.api.Test;
import tools.jackson.databind.JsonNode;
import tools.jackson.databind.json.JsonMapper;

/**
 * Cross-language golden-fixture round-trip on the Java side: every fixture under
 * {@code docs/contracts/fixtures/} is deserialized into its typed {@code core} record, re-serialized,
 * and the result is asserted <em>tree-equal</em> to the original fixture. The identical files are
 * round-tripped by the Go mirror ({@code client/internal/core}); if both suites are green the two
 * languages agree on the wire contract (issue #3 acceptance criterion 2).
 *
 * <p>Comparison is semantic (parsed {@link JsonNode} equality), not byte equality, so insignificant
 * whitespace in the human-edited fixtures does not matter — but field <em>names</em>, casing, types
 * and null-omission must match exactly, which is the property under test.
 *
 * <p>The fixtures live at the repo root, four levels up from {@code server/} where tests run; the
 * path is resolved relative to the module's working directory.
 */
class JsonRoundTripTest {

    /** Mapper mirroring the server's wire settings; see docs/contracts/serialization.md. */
    private static final JsonMapper MAPPER = JsonMapper.builder().build();

    private static final Path FIXTURES =
            Path.of("..", "docs", "contracts", "fixtures").toAbsolutePath().normalize();

    private static JsonNode readFixtureTree(String name) throws IOException {
        Path file = FIXTURES.resolve(name);
        assertThat(file).as("fixture %s exists", name).exists();
        return MAPPER.readTree(Files.readString(file));
    }

    /** Deserialize → re-serialize → assert tree-equal to the original fixture. */
    private <T> void assertRoundTrips(String fixtureName, Class<T> type) throws IOException {
        JsonNode original = readFixtureTree(fixtureName);
        T value = MAPPER.treeToValue(original, type);
        JsonNode reserialized = MAPPER.valueToTree(value);
        assertThat(reserialized)
                .as("round-trip of %s via %s", fixtureName, type.getSimpleName())
                .isEqualTo(original);
    }

    @Test
    void identityPageRoundTrips() throws IOException {
        assertRoundTrips("identity_page.json", Identity.class);
    }

    @Test
    void identityProjectRoundTrips() throws IOException {
        assertRoundTrips("identity_project.json", Identity.class);
    }

    @Test
    void pageRoundTrips() throws IOException {
        assertRoundTrips("page.json", Page.class);
    }

    @Test
    void observationRoundTrips() throws IOException {
        assertRoundTrips("observation.json", Observation.class);
    }

    @Test
    void observationMinimalRoundTrips() throws IOException {
        assertRoundTrips("observation_minimal.json", Observation.class);
    }

    @Test
    void sessionRoundTrips() throws IOException {
        assertRoundTrips("session.json", Session.class);
    }

    @Test
    void sessionOpenRoundTrips() throws IOException {
        assertRoundTrips("session_open.json", Session.class);
    }

    @Test
    void linkRoundTrips() throws IOException {
        assertRoundTrips("link.json", Link.class);
    }

    @Test
    void linkDeferredRoundTrips() throws IOException {
        assertRoundTrips("link_deferred.json", Link.class);
    }

    @Test
    void handoffRoundTrips() throws IOException {
        assertRoundTrips("handoff.json", Handoff.class);
    }

    @Test
    void handoffOpenRoundTrips() throws IOException {
        assertRoundTrips("handoff_open.json", Handoff.class);
    }

    @Test
    void observationKindsFixtureMatchesEnum() throws IOException {
        JsonNode tree = readFixtureTree("observation_kinds.json");
        @SuppressWarnings("unchecked")
        List<String> tokens = (List<String>) MAPPER.treeToValue(tree, List.class);

        // Every canonical kind appears in the fixture, in declaration order, and each token parses
        // back to the same constant: the enum and the contract fixture are locked together.
        assertThat(tokens).hasSameSizeAs(ObservationKind.values());
        for (int i = 0; i < tokens.size(); i++) {
            ObservationKind kind = ObservationKind.values()[i];
            assertThat(tokens.get(i)).isEqualTo(kind.wire());
            assertThat(ObservationKind.fromWire(tokens.get(i))).isEqualTo(kind);
        }
    }

    @Test
    void semanticEqualityIsOrderIndependentSanityCheck() throws IOException {
        // Guards the test's own premise: JsonNode equality ignores object key order, so a
        // re-serialization that differs only in field order would still (correctly) pass. Two
        // trees with the same entries in different source order must compare equal.
        JsonNode a = MAPPER.readTree("{\"workspace\":\"w\",\"project\":\"p\"}");
        JsonNode b = MAPPER.readTree("{\"project\":\"p\",\"workspace\":\"w\"}");
        assertThat(a).isEqualTo(b);
    }
}
